File: ADIVINHAR PALAVRAS/java/AdivinharPalavras/src/projeto/Palavra.java
package projeto;

/**
 * @author Tiago Penha Pedroso
 * Date: 05/05/2018, 20:28:22
 */

public class Palavra {
    
    //Atributos da classe
    public String texto;
    public int pontos;
    public String[] dicas;
    
    //Atributos Auxiliares
    private int contDicaAtual;
    
    public Palavra(String texto, String[] dicas){
        this.texto = texto.trim().toUpperCase();                       
        pontos = this.texto.length();
        this.dicas = dicas;
    }
    
    //alt + insert

    @Override
    public String toString() {
        String textoFinal = "";
        
        for (String dica : dicas) {
            textoFinal += "\n     \"" + dica + "\"";
        }
        
        return  "Palavra{\n" + 
                "  texto:  " + texto + "\n" +
                "  pontos: " + pontos + "\n" +
                "  dicas:  " + textoFinal + "\n" +
                "}";
    }

    @Override
    public boolean equals(Object outraPalavra) {
        if (outraPalavra instanceof Palavra) {
            Palavra outra = (Palavra) outraPalavra;            
            if (texto.equals(outra.texto)) {
                return true;
            }
        }
        
        return false;
    }
    
    public int getIndiceDicaAtual(){
        return contDicaAtual;
    }
    
    public int getLengthOfDicas(){
        return dicas.length;
    }
    
    public String getDicaAtual(){
        if (contDicaAtual < dicas.length) {
            return dicas[contDicaAtual];
        }

        return "";
    }
    
    public String getNextDica(){        
        String dicaAtual = getDicaAtual();
        contDicaAtual++;
        return dicaAtual;
    }
    
}

File: ADIVINHAR PALAVRAS/c#/AdivinharPalavras/Principal/Palavra.cs
using System;
using System.Collections.Generic;
using System.Text;


public class Palavra
{

    #region Atributos e Propriedades
    //Atributos da Classe
    public string texto;
    public int pontos;
    public string[] dicas;

    //Atributos Auxiliares
    private int indiceDicaAtual;

    //Constantes
    public static readonly Palavra VAZIA = new Palavra("", new string[] { "" });
    #endregion

    #region Construtores
    public Palavra(string texto, string[] dicas)
    {
        this.texto = texto.Trim().ToUpper();
        pontos = this.texto.Length;
        this.dicas = dicas;
    }
    #endregion

    #region Sobreescritas de Object e Sobrecargas de Operadores
    public override string ToString()
    {
        string mensagem = "Palavra{\n";
        mensagem += "   texto: " + texto + "\n";
        mensagem += "   pontos: " + pontos + "\n";
        mensagem += "   dicas: \n";

        foreach (var item in dicas)
        {
            mensagem += "     " + item + "\n";
        }

        mensagem += "}\n";
        return mensagem;

    }

    public override bool Equals(Object outraPalavra)
    {
        if (outraPalavra == null || !(outraPalavra is Palavra)) return false;
        //if (!(outraPalavra is Palavra) || outraPalavra == null) return false;

        Palavra outra = outraPalavra as Palavra;
        return texto == outra.texto;
    }

    public static bool operator ==(Palavra palavra1, Palavra palavra2)
    {
        if (Object.Equals(palavra1, null) || Object.Equals(palavra2, null)) return false;

        return palavra1.Equals(palavra2);
    }

    public static bool operator !=(Palavra palavra1, Palavra palavra2)
    {
        return !(palavra1 == palavra2);
    }
    #endregion

    #region Outros Métodos
    public int GetIndiceDicaAtual()
    {
        return indiceDicaAtual;
    }

    public int GetLengthOfDicas()
    {
        return dicas.Length;
    }

    public string GetDicaAtual()
    {
        if (indiceDicaAtual < dicas.Length)
        {
            return dicas[indiceDicaAtual];
        }

        return "";
    }

    public string GetNextDica()
    {
        string dicaAtual = GetDicaAtual();
        indiceDicaAtual++;
        return dicaAtual;
    }
    #endregion

}